File: leetcode4/GCDofOddandEvenSums/a.py
# -*- coding: utf-8 -*-
# Copyright (C) 2025 by iamslash 

# 0ms 100.00%
# math
# O(1) O(1)
class Solution:
    def gcd(self, a: int, b: int) -> int:
        # base
        if b == 0:
            return a        
        # recursion
        return gcd(b, a % b)
    
    def gcdOfOddEvenSums(self, n: int) -> int:

        nn = n + n
        sum = nn * (nn + 1) // 2
        sum_odd = sum - n
        sum_evn = sum_odd + n

        return gcd(sum_evn, sum_odd)

File: leetcode4/GCDofOddandEvenSums/a.cpp
/* Copyright (C) 2025 by iamslash */

#include <cstdio>

// math
// O(1) O(1)
class Solution {

private:
  int gcd(int a, int b) {
    // base
    if (b == 0) {
      return a;
    }
    
    // recursion
    return gcd(b, a % b);
  }
  
public:
  int gcdOfOddEvenSums(int n) {
    int nn = n + n;
    int sum = nn * (nn + 1) / 2;

    int sum_odd = (sum - n);
    int sum_evn = sum_odd + n;

    return gcd(sum_evn, sum_odd);
  }
  
};

int main() {
  Solution sln;
  
  return sln.gcdOfOddEvenSums(10);
}
